### DISCUSSION TITLE
Teleport over ssh via identity file

### DISCUSSION THREAD

**Discussion by igtsekov** [gravitational/teleport#7278] [Q&A] [answered]
I found out that I can use identity files to log in to a node of the cluster, without login into the cluster.
I did these steps:

1. I create a pem file `tctl auth sign --ttl=30h --user=<user> --out=file.pem`
2. I copy it to the location that I want to ssh from.
3. I executed this command `tsh --proxy=<proxy>:<port> ssh -i file.pem <user>@<IP>`
4. Everything work as expected. I was successfully logged in to the node.

I want to optimize this to be used via ssh. So I created a `.ssh/config`
```
Host <IP>
  Port 3022
  ProxyCommand tsh --proxy=<proxy>:<port> ssh -i file.pem <user>@<IP>
```
In order to test it, but I get an error:
```
ssh <IP>
Bad packet length 218783322.
ssh_dispatch_run_fatal: Connection to UNKNOWN port 0: message authentication code incorrect
```
According to the [ssh ProxyCommand dokumentation](https://linux.die.net/man/5/ssh_config), this should be possible.
I am using teleport version 6.0.2 for the client and server.

Did someone try to do something like this? Is this even possible?

**Comment by webvictim** [comment, marked as answer]
`tsh ssh` isn't designed to be used directly as a `ProxyCommand` because it does more than just forward stdin/stdout (which is what a `ProxyCommand` expects). 

You should be able to use `ssh` itself directly with its proxy subsystem:

```
Host <IP>
  User <user>
  Port 3022
  IdentityFile file.pem
  ProxyCommand ssh <proxy> -s proxy:%h:%p

Host <proxy>
  User <user>
  Port 3023
  IdentityFile file.pem
```

At this point, `ssh <IP>` by itself on the command line should work and can be used by Ansible/other tools which expect an `ssh` transport.

**Reply by igtsekov** [reply]
According to the docs that [describe how to access the node via ssh client](https://goteleport.com/docs/server-access/guides/openssh/) I need to be logged in into the cluster. This makes the identity file pointless. If I am logged into the cluster I will get all the keys from there.
I want to access the nodes via teleport without executing the `tsh login` command. and use the identity file.
As I said earlier `tsh --proxy=<proxy>:<port> ssh -i file.pem <user>@<IP>` works when you are not login into the teleport cluster. I want to achieve this behaviour via ssh.
Is it possible?

**Reply by webvictim** [reply]
Did you try the SSH config I posted above? That should work without you needing to be logged into the cluster as it uses `IdentityFile file.pem` (which is the equivalent of `-i file.pem`)

**Reply by igtsekov** [reply]
Yes, I did. I am missing something because I get an error `Permission denied (publickey).` I will go over the doc files to see what I have missed.

**Reply by webvictim** [reply]
If you post the output of `ssh -vv hostname` I can try to help you fix it.

**Reply by igtsekov** [reply]
Here is the config
```
Host 192.168.1.1
  User jenkins
  Port 3022
  IdentityFile /var/lib/jenkins/jenkins.pem
  ProxyCommand ssh teleport -s proxy:%h:%p

Host teleport
  User jenkins
  Port 3023
  IdentityFile /var/lib/jenkins/jenkins.pem
```
Here is the output
```
ssh -vv 192.168.1.1
OpenSSH_7.2p2 Ubuntu-4ubuntu2.10, OpenSSL 1.0.2g  1 Mar 2016
debug1: Reading configuration data /var/lib/jenkins/.ssh/config
debug1: /var/lib/jenkins/.ssh/config line 1: Applying options for 192.168.1.1
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug1: Executing proxy command: exec ssh teleporta -s proxy:192.168.1.1:3022
debug1: key_load_public: No such file or directory
debug1: identity file /var/lib/jenkins/jenkins.pem type -1
debug1: permanently_drop_suid: 20054
debug1: key_load_public: No such file or directory
debug1: identity file /var/lib/jenkins/jenkins.pem-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.10
Permission denied (publickey).
ssh_exchange_identification: Connection closed by remote host
```

**Reply by webvictim** [reply]
This line looks wrong:
```exec ssh teleporta -s proxy:192.168.1.1:3022```

Is `teleporta` a typo? Something may be coming from `/etc/ssh/ssh_config`.

**Reply by igtsekov** [reply]
It is a typo when I copy/paste the output. I am not sure how I manage to do it. In the config is the correct value
```
Host teleporta
  User jenkins
  Port 3023
  IdentityFile /var/lib/jenkins/jenkins.pem
```

**Reply by igtsekov** [reply]
I tested it on a different server that has only
```
Host *
	SendEnv LANG LC_*
```
in the `ssh_config` and still, I get the same error.

**Reply by webvictim** [reply]
Can you take a look at the Teleport proxy server logs during the failed connection and see if there’s any output?

The `type -1` in the output generally indicates that the identity file cannot be loaded. What's the output of `ls -l /var/lib/jenkins/jenkins.pem`?

**Reply by igtsekov** [reply]
The pem file:
```
ls -l /var/lib/jenkins/jenkins.pem
-rw------- 1 jenkins default 6593 Jun 17 10:17 /var/lib/jenkins/jenkins.pem
```
This is the DEBUG log from the proxy
```
DEBU [PROXY]     conn(<CLIENT-IP>:60942-><PORXY-IP>:3023, user=jenkins) auth attempt fingerprint:ssh-rsa SHA256:jgcOTNZ2smq7HCQHrMNrER3d68ouX5RnBznJrjaXjIY local:<PORXY-IP>:3023 remote:<CLIENT-IP>:60942 user:jenkins srv/authhandlers.go:163
DEBU [PROXY]     conn(<CLIENT-IP>:60942-><PORXY-IP>:3023, user=jenkins) auth attempt with key ssh-rsa SHA256:jgcOTNZ2smq7HCQHrMNrER3d68ouX5RnBznJrjaXjIY, (*ssh.Certificate)(nil) fingerprint:ssh-rsa SHA256:jgcOTNZ2smq7HCQHrMNrER3d68ouX5RnBznJrjaXjIY local:<PORXY-IP>:3023 remote:<CLIENT-IP>:60942 user:jenkins srv/authhandlers.go:166
DEBU [PROXY]     auth attempt, unsupported key type fingerprint:ssh-rsa SHA256:jgcOTNZ2smq7HCQHrMNrER3d68ouX5RnBznJrjaXjIY local:<PORXY-IP>:3023 remote:<CLIENT-IP>:60942 user:jenkins srv/authhandlers.go:168
```
I will check to see why the pem file is not loaded.

**Reply by webvictim** [reply]
What's the version of `sshd` running on the machine that you're connecting to? You can check with `ssh -V`. If it's 8.2 or newer then the `ssh-rsa` certificate algorithm may have been deprecated.

**Reply by igtsekov** [reply]
We are using `OpenSSH_7.2p2` as client and server.

**Reply by webvictim** [reply]
Looks like maybe Jenkins is starting an SSH agent and refusing to use the key/certificate provided? You can add `IdentitiesOnly yes` to both host sections in your `ssh_config` to disable the use of any ssh-agent identity and force the use of identity files only.

**Reply by igtsekov** [reply]
Still got the same error message.

**Reply by webvictim** [reply]
I've just noticed - the regular .pem files from `tctl auth sign` are Teleport-specific bundles that the SSH client itself cannot understand. Try exporting the SSH cert using `--format openssh`, like this:

`tctl auth sign --ttl=30h --user=jenkins --format=openssh --out=file.pem`

You should get two files generated - `file.pem` and `file.pem-cert` - these will both need to be copied to your `/var/lib/jenkins` directory. After this it should work with the SSH client as expected.

**Reply by igtsekov** [reply]
I did it. But still got the same error. SSH was able to load the `jenkins.pem-cert` file and asked me for a password.
```
ssh 1192.168.1.1 -vvv
OpenSSH_7.2p2 Ubuntu-4ubuntu2.10, OpenSSL 1.0.2g  1 Mar 2016
debug1: Reading configuration data /var/lib/jenkins/.ssh/config
debug1: /var/lib/jenkins/.ssh/config line 1: Applying options for 192.168.1.1
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug1: Executing proxy command: exec ssh teleporta -s proxy:192.168.1.1:3022
debug1: key_load_public: No such file or directory
debug1: identity file /var/lib/jenkins/jenkins.pem type -1
debug1: permanently_drop_suid: 20054
debug1: identity file /var/lib/jenkins/jenkins.pem-cert type 5
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.10
Enter passphrase for key '/var/lib/jenkins/jenkins.pem-cert':
Permission denied (publickey).
ssh_exchange_identification: Connection closed by remote host
```

**Reply by webvictim** [reply]
There should be no passphrase on that file. 

Also `debug1: identity file /var/lib/jenkins/jenkins.pem type -1` looks like an error loading that file. Can you show `ls -l /var/lib/jenkins` and double check that you copied both generated files from `tctl auth sign` correctly?

**Reply by igtsekov** [reply]
```
-rw-------   1 jenkins default    1675 Jun 25 13:42 jenkins.pem
-rw-------   1 jenkins default    1654 Jun 25 13:42 jenkins.pem-cert.pub
```
The files are the same. I double and triple check them.
Yes, I think that this `debug1: identity file /var/lib/jenkins/jenkins.pem type -1` is the problem. I am searching for how to solve this, but no success.

**Reply by webvictim** [reply]
Here's what the output looks like for me locally (with a successful SSH):

```
debug1: identity file /Users/gus/.tsh/keys/teleport.example.com/webvictim type 0
debug1: identity file /Users/gus/.tsh/keys/teleport.example.com/webvictim-cert type -1
```

If you haven't seen https://github.com/gravitational/teleport/discussions/5530 it may be helpful, as it discusses all the steps involved.

How are you providing the key files to Jenkins? Are you loading them directly onto the Jenkins server via `scp` or similar, or have you added them as secrets which are provided to the build job?

**Reply by igtsekov** [reply]
I used `scp` and manual copy/paste it. In both cases I get `debug1: identity file /var/lib/jenkins/jenkins.pem type -1`

I went over the example provided and I see that it is login into the cluster `It will check whether my Teleport certificates have expired and if they have, it will run tsh login to reauthenticate.` As I mentioned before - this is not an option. I need to access a node without login into the teleport cluster. 

If we ignore this part (auto logger) all other configs are the same.

**Reply by igtsekov** [reply]
After some additional try and error and some strace I was able to fix the `debug1: identity file /var/lib/jenkins/jenkins.pem type -1` error. 
It turns out that I need these set to files
```
-rw------- 1 jenkins default 1679 Jul 12 10:37 .ssh/jenkins.pem
-rw------- 1 jenkins default 1654 Jul 12 11:36 .ssh/jenkins.pem-cert
-rw------- 1 jenkins default 1654 Jul 12 11:31 .ssh/jenkins.pem.pub
```
Where `jenkins.pem-cert` and `jenkins.pem.pub` are the same file.
Now I get new error
```
agent forwarding not requested or not authorizedsubsystem request failed on channel 0
ssh_exchange_identification: Connection closed by remote host
```
I will continue debugging it.

**Reply by webvictim** [reply]
Is your Teleport cluster running in recording proxy mode? If so, your SSH commands will need to use `ssh -A` instead (or set `ForwardAgent yes` in your SSH config file) so that SSH agent forwarding is enabled and Teleport can intercept the SSH traffic.

Your Teleport role will also need to allow agent forwarding:

```yaml
spec:
  options:
    forward_agent: true
```

I'm not sure how you ended up needing to rename files, but I'm pleased that it works now. Usually when you run `tctl auth sign --format=openssh --out=jenkins.pem` it should generate both `jenkins.pem` and `jenkins.pem-cert.pub` for you:

```console
root@artemis:~/ssh# tctl auth sign --format=openssh --user=gus --ttl=5m --out=jenkins.pem
The credentials have been written to jenkins.pem, jenkins.pem-cert.pub

root@artemis:~/ssh# ls -l
total 8
-rw------- 1 root root 1675 Jul 12 13:32 jenkins.pem
-rw------- 1 root root 1782 Jul 12 13:32 jenkins.pem-cert.pub
```

You can specify a different location to the certificate file (`-cert.pub`) using the `CertificateFile` directive in the SSH config if needed:

```
Host hostname
    User gus
    Port 3022
    IdentityFile /var/lib/jenkins/jenkins.pem
    CertificateFile /var/lib/jenkins/jenkins.pem-cert.pub
    IdentitiesOnly yes
    ForwardAgent yes
```

By default, SSH takes the name of the `IdentityFile` (`jenkins.pem`) and just appends `-cert.pub` to the end to find the `CertificateFile`.

**Reply by igtsekov** [reply]
I needed to add the key to the ssh-agent, so that the `ForwardAgent` option can forward it. I was able to reach the node but the connection was closed automatically.
Now I get a different error :( from the teleport node logs:
```
DEBU [NODE]      conn(<PROXY-IP>:36634-><NODE-IP>:3022, user=jenkins) auth attempt fingerprint:ssh-rsa SHA256:GCR3sEmz171CVsxqmKPJwlUJZth9k15yTQ7rQuE+sYQ local:<NODE-IP>:3022 remote:<PROXY-IP>:36634 user:jenkins srv/authhandlers.go:163
DEBU [NODE]      conn(<PROXY-IP>:36634-><NODE-IP>:3022, user=jenkins) auth attempt with key ssh-rsa SHA256:GCR3sEmz171CVsxqmKPJwlUJZth9k15yTQ7rQuE+sYQ, (*ssh.Certificate)(nil) fingerprint:ssh-rsa SHA256:GCR3sEmz171CVsxqmKPJwlUJZth9k15yTQ7rQuE+sYQ local:<NODE-IP>:3022 remote:<PROXY-IP>:36634 user:jenkins srv/authhandlers.go:166
DEBU [NODE]      auth attempt, unsupported key type fingerprint:ssh-rsa SHA256:GCR3sEmz171CVsxqmKPJwlUJZth9k15yTQ7rQuE+sYQ local:<NODE-IP>:3022 remote:<PROXY-IP>36634 user:jenkins srv/authhandlers.go:168
```
I am using version 6.0.2 of teleport.

**Reply by webvictim** [reply]
It looks to me like it's failing to pass the certificate for some reason. What's the output of `ssh-add -L` after the key is added to the agent? (You can post the whole output, it's only public keys)

Can you SSH manually to the proxy using the key from the agent with agent forwarding enabled? What does `ssh-add -L` show once you have a session on the proxy?

**Reply by igtsekov** [reply]
On the Jenkins node
```
ssh-add -L
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDYevSq4O7BrZ9xhaJHBzo4dCXbxaJamqouGkd1pIeZFgVs87/4nNh4CiUY+VmGoztkNk2KsDh9spgGvdbNOGAGXhNktrdRHL/R+lzx3hMfrHt39BCRWFIcDTRebt1rzhL2ihPsANaDdW9tHEUoCGoqlcSQT8CpwkHFqvf2ZJZ8RJExe5lvw0LKJIkGS9LvTj3xNWZjjkN8rgRkj830UlItqZ1iQCOQnP++tjpVIK3lN4auAdMzpvv29wv1uVlLXHPvDWlJMT5rMgtwQJS/xB0xqXAUD2xnenR13DqVWmIQF83utqs9etXsYixMiRaxApWMNyPMtcZadkUND0J9VIeJ jenkins.pem
```
After that, I ssh to the proxy node
```
ssh -p 3022 -A <PROXY-IP> -i jenkins.pem
```
On the proxy node
```
ssh-add -L
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDYevSq4O7BrZ9xhaJHBzo4dCXbxaJamqouGkd1pIeZFgVs87/4nNh4CiUY+VmGoztkNk2KsDh9spgGvdbNOGAGXhNktrdRHL/R+lzx3hMfrHt39BCRWFIcDTRebt1rzhL2ihPsANaDdW9tHEUoCGoqlcSQT8CpwkHFqvf2ZJZ8RJExe5lvw0LKJIkGS9LvTj3xNWZjjkN8rgRkj830UlItqZ1iQCOQnP++tjpVIK3lN4auAdMzpvv29wv1uVlLXHPvDWlJMT5rMgtwQJS/xB0xqXAUD2xnenR13DqVWmIQF83utqs9etXsYixMiRaxApWMNyPMtcZadkUND0J9VIeJ jenkins.pem
```
It seems that the key is forwarded correctly to the proxy.

**Reply by webvictim** [reply]
The key is forwarded correctly, but there is no SSH certificate present in either listing. You should see two entries - one for the key and one for the certificate.

It seems that `jenkins.pem-cert.pub` is still not being added to the SSH agent. How are you adding the key file?

Can you run `ssh-keygen -Lf jenkins.pem-cert.pub` and provide the output?

**Reply by igtsekov** [reply]
It works. I have ssh access via da proxy node. THANK YOU very much for the step by step help.

**Reply by webvictim** [reply]
You're very welcome, I'm glad we got there in the end.

For the benefit of the thread and anyone who may read this in future, could you please explain what the issue was in the end and how you fixed it?

**Reply by igtsekov** [reply]
To summarize it all
 I needed to add `--format=openssh` to the `tctl auth sign` command. 
When importing the keys into the ssh-agent I renamed the cert file by mistake, because of that it was not imported.
I needed to update the public cluster key that I generated with this command `sudo tctl auth export --type=host > teleport_host_ca.pub`. The change was that I substitute the cluster name with the IP range of the instances (`192.*.*.*`) and  🎉
